ligo/stablecoin/operator.ligo

```diff
@@ -0,0 +1,158 @@
+ * TODO: we should probably ommit `owner` in `operator_param_` as @gromak once suggested
+ *)
+function validate_operator_owner_is_sender
+  ( const param : operator_param
+  ) : unit is if param.0 = Tezos.sender
+  then unit
+  else failwith ("valdiate_oeprator_param: Sender is not token owner")
+
+(*
+ * Checks whether the given operator is present in provided operator map
+ *)
+function is_operator_impl
+  ( const param     : operator_param
+  ; const operators : operators
+  ) : bool is block
+  { const operator_key : (owner * operator) = (param.0, param.1.0)
+  } with case Big_map.find_opt (operator_key, operators) of
+      None -> False
+    | Some (u) -> True
+    end
+
+(*
+ * Returns a callback of `Is_operator` entrypoint call
+ *)
+function is_operator
+  ( const param     : is_operator_params
+  ; const operators : operators
+  ) : operation is block
+{ const response : is_operator_response =
+    (param.0, is_operator_impl (param.0, operators))
+} with Tezos.transaction (response, 0mutez, param.1)
+
+(*
+ * Validates operators for the given tranfser batch
+ * and operator storage
+ *)
+function validate_operators
+  ( const permissions_descriptor : permissions_descriptor_
+  ; const transfer_parameters    : transfer_params
+  ; const operators              : operators
+  ) : unit is block
+{ const self_transfer_policy_ : self_transfer_policy_ =
+    Layout.convert_from_right_comb ((permissions_descriptor.self : self_transfer_policy))
+; const can_transfer_to_self : bool =
+    case self_transfer_policy_ of
+      Self_transfer_permitted (u) -> True
+    | Self_transfer_denied    (u) -> False
+    end
+
+; const operator_transfer_policy_ : operator_transfer_policy_ =
+    Layout.convert_from_right_comb ((permissions_descriptor.operator : operator_transfer_policy))
+; const can_transfer_operator : bool =
+    case operator_transfer_policy_ of
+      Operator_transfer_permitted (u) -> True
+    | Operator_transfer_denied    (u) -> False
+    end
+
+; const operator : address = Tezos.sender
+
+; function owner_folding
+    ( const owner_tokens_accumulator : owner_tokens
+    ; const transfer_operation       : transfer_param
+    ) : owner_tokens is block
+  { const tokens : option (set (token_id)) =
+      Map.find_opt (transfer_operation.0, owner_tokens_accumulator)
+  ; const updated_tokens : set (token_id) =
+      case tokens of
+        None -> set [transfer_operation.1.1.0]
+      | Some (tokens) -> Set.add (transfer_operation.1.1.0, tokens)
+      end
+  } with Map.update
+      ( transfer_operation.0
+      , Some (updated_tokens)
+      , owner_tokens_accumulator
+      )
+
+; const owner_tokens : owner_tokens = List.fold
+    ( owner_folding
+    , transfer_parameters
+    , (map [] : owner_tokens)
+    )
+
+; function validate_owner
+    ( const owner  : address
+    ; const tokens : set (token_id)
+    ) : unit is block
+  { var is_operator : bool := False
+  ; if can_transfer_to_self and owner = operator
+    then skip
```

Comment:
Is this a bug here? if both `can_transfer_to_self` and `owner = operator` is true, then `is_operator` will remain false. Right?

Reply:
Yep, refresh your page please